File: wrappers/java/foundation/src/main/java/com/virgilsecurity/crypto/foundation/Pkcs8Serializer.java
/*
* Copyright (C) 2015-2020 Virgil Security, Inc.
*
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* (1) Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* (2) Redistributions in binary form must reproduce the above copyright
* notice, this list of conditions and the following disclaimer in
* the documentation and/or other materials provided with the
* distribution.
*
* (3) Neither the name of the copyright holder nor the names of its
* contributors may be used to endorse or promote products derived from
* this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
* STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
* IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*
* Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
*/

package com.virgilsecurity.crypto.foundation;

/*
* Implements PKCS#8 key serialization to DER format.
*/
public class Pkcs8Serializer implements AutoCloseable, KeySerializer {

    public long cCtx;

    /* Create underlying C context. */
    public Pkcs8Serializer() {
        super();
        this.cCtx = FoundationJNI.INSTANCE.pkcs8Serializer_new();
    }

    /* Wrap underlying C context. */
    Pkcs8Serializer(FoundationContextHolder contextHolder) {
        this.cCtx = contextHolder.cCtx;
    }

    public void setAsn1Writer(Asn1Writer asn1Writer) {
        FoundationJNI.INSTANCE.pkcs8Serializer_setAsn1Writer(this.cCtx, asn1Writer);
    }

    /*
    * Setup predefined values to the uninitialized class dependencies.
    */
    public void setupDefaults() {
        FoundationJNI.INSTANCE.pkcs8Serializer_setupDefaults(this.cCtx);
    }

    /*
    * Serialize Public Key by using internal ASN.1 writer.
    * Note, that caller code is responsible to reset ASN.1 writer with
    * an output buffer.
    */
    public int serializePublicKeyInplace(RawPublicKey publicKey) throws FoundationException {
        return FoundationJNI.INSTANCE.pkcs8Serializer_serializePublicKeyInplace(this.cCtx, publicKey);
    }

    /*
    * Serialize Private Key by using internal ASN.1 writer.
    * Note, that caller code is responsible to reset ASN.1 writer with
    * an output buffer.
    */
    public int serializePrivateKeyInplace(RawPrivateKey privateKey) throws FoundationException {
        return FoundationJNI.INSTANCE.pkcs8Serializer_serializePrivateKeyInplace(this.cCtx, privateKey);
    }

    /*
    * Acquire C context.
    * Note. This method is used in generated code only, and SHOULD NOT be used in another way.
    */
    public static Pkcs8Serializer getInstance(long cCtx) {
        FoundationContextHolder ctxHolder = new FoundationContextHolder(cCtx);
        return new Pkcs8Serializer(ctxHolder);
    }

    /* Close resource. */
    public void close() {
        FoundationJNI.INSTANCE.pkcs8Serializer_close(this.cCtx);
    }

    /*
    * Calculate buffer size enough to hold serialized public key.
    *
    * Precondition: public key must be exportable.
    */
    public int serializedPublicKeyLen(RawPublicKey publicKey) {
        return FoundationJNI.INSTANCE.pkcs8Serializer_serializedPublicKeyLen(this.cCtx, publicKey);
    }

    /*
    * Serialize given public key to an interchangeable format.
    *
    * Precondition: public key must be exportable.
    */
    public byte[] serializePublicKey(RawPublicKey publicKey) throws FoundationException {
        return FoundationJNI.INSTANCE.pkcs8Serializer_serializePublicKey(this.cCtx, publicKey);
    }

    /*
    * Calculate buffer size enough to hold serialized private key.
    *
    * Precondition: private key must be exportable.
    */
    public int serializedPrivateKeyLen(RawPrivateKey privateKey) {
        return FoundationJNI.INSTANCE.pkcs8Serializer_serializedPrivateKeyLen(this.cCtx, privateKey);
    }

    /*
    * Serialize given private key to an interchangeable format.
    *
    * Precondition: private key must be exportable.
    */
    public byte[] serializePrivateKey(RawPrivateKey privateKey) throws FoundationException {
        return FoundationJNI.INSTANCE.pkcs8Serializer_serializePrivateKey(this.cCtx, privateKey);
    }
}

File: wrappers/wasm/foundation/src/Pkcs8Serializer.js
/**
 * Copyright (C) 2015-2022 Virgil Security, Inc.
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * (1) Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * (2) Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in
 * the documentation and/or other materials provided with the
 * distribution.
 *
 * (3) Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 */


const precondition = require('./precondition');

const initPkcs8Serializer = (Module, modules) => {
    class Pkcs8Serializer {

        constructor(ctxPtr) {
            this.name = 'Pkcs8Serializer';

            if (typeof ctxPtr === 'undefined') {
                this.ctxPtr = Module._vscf_pkcs8_serializer_new();
            } else {
                this.ctxPtr = ctxPtr;
            }
        }

        static newAndUseCContext(ctxPtr) {
            // assert(typeof ctxPtr === 'number');
            return new Pkcs8Serializer(Module._vscf_pkcs8_serializer_shallow_copy(ctxPtr));
        }

        static newAndTakeCContext(ctxPtr) {
            // assert(typeof ctxPtr === 'number');
            return new Pkcs8Serializer(ctxPtr);
        }

        delete() {
            if (typeof this.ctxPtr !== 'undefined' && this.ctxPtr !== null) {
                Module._vscf_pkcs8_serializer_delete(this.ctxPtr);
                this.ctxPtr = null;
            }
        }

        asn1Writer(asn1Writer) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureImplementInterface('asn1Writer', asn1Writer, 'Foundation.Asn1Writer', modules.FoundationInterfaceTag.ASN1_WRITER, modules.FoundationInterface);
            Module._vscf_pkcs8_serializer_release_asn1_writer(this.ctxPtr)
            Module._vscf_pkcs8_serializer_use_asn1_writer(this.ctxPtr, asn1Writer.ctxPtr)
        }

        serializedPublicKeyLen(publicKey) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureClass('publicKey', publicKey, modules.RawPublicKey);
            
            let proxyResult;
            proxyResult = Module._vscf_pkcs8_serializer_serialized_public_key_len(this.ctxPtr, publicKey.ctxPtr);
            return proxyResult;
        }

        serializePublicKey(publicKey) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureClass('publicKey', publicKey, modules.RawPublicKey);
            
            const outCapacity = this.serializedPublicKeyLen(publicKey);
            const outCtxPtr = Module._vsc_buffer_new_with_capacity(outCapacity);
            
            try {
                const proxyResult = Module._vscf_pkcs8_serializer_serialize_public_key(this.ctxPtr, publicKey.ctxPtr, outCtxPtr);
                modules.FoundationError.handleStatusCode(proxyResult);
            
                const outPtr = Module._vsc_buffer_bytes(outCtxPtr);
                const outPtrLen = Module._vsc_buffer_len(outCtxPtr);
                const out = Module.HEAPU8.slice(outPtr, outPtr + outPtrLen);
                return out;
            } finally {
                Module._vsc_buffer_delete(outCtxPtr);
            }
        }

        serializedPrivateKeyLen(privateKey) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureClass('privateKey', privateKey, modules.RawPrivateKey);
            
            let proxyResult;
            proxyResult = Module._vscf_pkcs8_serializer_serialized_private_key_len(this.ctxPtr, privateKey.ctxPtr);
            return proxyResult;
        }

        serializePrivateKey(privateKey) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureClass('privateKey', privateKey, modules.RawPrivateKey);
            
            const outCapacity = this.serializedPrivateKeyLen(privateKey);
            const outCtxPtr = Module._vsc_buffer_new_with_capacity(outCapacity);
            
            try {
                const proxyResult = Module._vscf_pkcs8_serializer_serialize_private_key(this.ctxPtr, privateKey.ctxPtr, outCtxPtr);
                modules.FoundationError.handleStatusCode(proxyResult);
            
                const outPtr = Module._vsc_buffer_bytes(outCtxPtr);
                const outPtrLen = Module._vsc_buffer_len(outCtxPtr);
                const out = Module.HEAPU8.slice(outPtr, outPtr + outPtrLen);
                return out;
            } finally {
                Module._vsc_buffer_delete(outCtxPtr);
            }
        }

        setupDefaults() {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            Module._vscf_pkcs8_serializer_setup_defaults(this.ctxPtr);
        }

        serializePublicKeyInplace(publicKey) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureClass('publicKey', publicKey, modules.RawPublicKey);
            const proxyResult = Module._vscf_pkcs8_serializer_serialize_public_key_inplace(this.ctxPtr, publicKey.ctxPtr);
            modules.FoundationError.handleStatusCode(proxyResult);
        }

        serializePrivateKeyInplace(privateKey) {
            precondition.ensureNotNull('this.ctxPtr', this.ctxPtr);
            precondition.ensureClass('privateKey', privateKey, modules.RawPrivateKey);
            const proxyResult = Module._vscf_pkcs8_serializer_serialize_private_key_inplace(this.ctxPtr, privateKey.ctxPtr);
            modules.FoundationError.handleStatusCode(proxyResult);
        }

    }

    return Pkcs8Serializer;
};

module.exports = initPkcs8Serializer;
